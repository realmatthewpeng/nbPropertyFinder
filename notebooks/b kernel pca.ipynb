{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# import library\n",
    "import tensorflow as tf\n",
    "import numpy as np\n",
    "import sys, os,cv2\n",
    "from scipy.misc import imread,imresize\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from imgaug import augmenters as iaa\n",
    "import imgaug as ia\n",
    "from scipy.ndimage import zoom\n",
    "from sklearn.utils import shuffle\n",
    "import matplotlib.animation as animation\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from sklearn.datasets import make_blobs,make_moons,make_swiss_roll\n",
    "\n",
    "plt.style.use('seaborn')\n",
    "os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# load some data set\n",
    "n_samples = 4000\n",
    "X, y = make_moons(n_samples=n_samples, random_state=123,noise=0.1)\n",
    "X    = X - X.mean(0,keepdims=True)\n",
    "plt.figure(figsize=(8,6))\n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "plt.title('Current Data Set'); plt.ylabel('y coordinate'); plt.xlabel('x coordinate'); plt.grid(True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Do EVD to see the projection matrix\n",
    "cov        = X.T @ X /(X.shape[0] - 1)\n",
    "val,vec    = np.linalg.eigh(cov)\n",
    "idx        = np.argsort(val)[::-1]; val = val[idx]; vec = vec[:,idx]\n",
    "project_X  = X   @ vec;                      project_V  = vec.T @ vec\n",
    "revert_X   = project_X @ np.linalg.inv(vec) ;revertedV  = project_V @ np.linalg.inv(vec).T\n",
    "\n",
    "plt.figure(figsize=(15,5)); \n",
    "\n",
    "plt.subplot(131); \n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(132); \n",
    "plt.scatter(project_X[y==0, 0], project_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(project_X[y==1, 0], project_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *project_V[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *project_V[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(133); \n",
    "plt.scatter(revert_X[y==0, 0], revert_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(revert_X[y==1, 0], revert_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *revertedV[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *revertedV[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# perform D reduction\n",
    "cov        = X.T @ X /(X.shape[0] - 1)\n",
    "val,vec    = np.linalg.eigh(cov)\n",
    "idx        = np.argsort(val)[::-1]; val = val[idx]; vec = vec[:,idx]\n",
    "\n",
    "vec_reduced= np.zeros_like(vec)\n",
    "vec_reduced[:,:1] = vec[:,:1]\n",
    "val_reduced= val.copy()\n",
    "val_reduced[-1:]= 0\n",
    "\n",
    "project_X  = X   @ vec_reduced;                      project_V  = vec_reduced.T @ vec_reduced\n",
    "revert_X   = project_X @ np.linalg.inv(vec_reduced+0.0001) ;revertedV  = project_V.T @ np.linalg.inv(vec_reduced+0.0001)\n",
    "\n",
    "plt.figure(figsize=(15,5)); \n",
    "\n",
    "plt.subplot(131); \n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(132); \n",
    "plt.scatter(project_X[y==0, 0], project_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(project_X[y==1, 0], project_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *project_V[:,0] * val_reduced[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *project_V[:,1] * val_reduced[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(133); \n",
    "plt.scatter(revert_X[y==0, 0], revert_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(revert_X[y==1, 0], revert_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *revertedV[:,0] * val_reduced[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *revertedV[:,1] * val_reduced[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     12
    ]
   },
   "outputs": [],
   "source": [
    "# do the reduction but in a different way\n",
    "cov        = X.T @ X /(X.shape[0] - 1)\n",
    "val,vec    = np.linalg.eigh(cov)\n",
    "idx        = np.argsort(val)[::-1]; val = val[idx]; vec = vec[:,idx]\n",
    "project_X    = X   @ vec;                      project_V  = vec.T @ vec\n",
    "\n",
    "inversed_vec  = np.linalg.inv(vec)\n",
    "inversed_vec  = inversed_vec - inversed_vec.mean(1)\n",
    "revert_X      = project_X @ inversed_vec ;revertedV  = project_V @ inversed_vec.T\n",
    "\n",
    "\n",
    "def plot():\n",
    "    plt.figure(figsize=(15,5)); \n",
    "\n",
    "    plt.subplot(131); \n",
    "    plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "    plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "    pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "    pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "    plt.grid(True); \n",
    "\n",
    "    plt.subplot(132); \n",
    "    plt.scatter(project_X[y==0, 0], project_X[y==0, 1], color='red',  alpha=0.5)\n",
    "    plt.scatter(project_X[y==1, 0], project_X[y==1, 1], color='blue', alpha=0.5)\n",
    "    pca21=plt.arrow   (0, 0, *project_V[:,0] * val_reduced[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "    pca22=plt.arrow   (0, 0, *project_V[:,1] * val_reduced[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "    plt.grid(True); \n",
    "\n",
    "    plt.subplot(133); \n",
    "    plt.scatter(revert_X[y==0, 0], revert_X[y==0, 1], color='red',  alpha=0.5)\n",
    "    plt.scatter(revert_X[y==1, 0], revert_X[y==1, 1], color='blue', alpha=0.5)\n",
    "    pca21=plt.arrow   (0, 0, *revertedV[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "    pca22=plt.arrow   (0, 0, *revertedV[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "    plt.grid(True); \n",
    "\n",
    "    plt.show()\n",
    "plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do EVD to see the projection matrix\n",
    "cov        = X.T @ X /(X.shape[0] - 1)\n",
    "val,vec    = np.linalg.eigh(cov)\n",
    "idx        = np.argsort(val)[::-1]; val = val[idx]; vec = vec[:,idx]\n",
    "project_X  = X   @ vec;                      project_V  = vec.T @ vec\n",
    "\n",
    "plt.figure(figsize=(15,5)); \n",
    "\n",
    "plt.subplot(141); \n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(142); \n",
    "plt.scatter(project_X[y==0, 0], project_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(project_X[y==1, 0], project_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *project_V[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *project_V[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "inverse  = np.linalg.inv(vec)\n",
    "inverse  = inverse - inverse.mean(0,keepdims=True) \n",
    "revert_X   = project_X @ inverse ;revertedV  = project_V @ inverse.T\n",
    "\n",
    "plt.subplot(143); \n",
    "plt.scatter(revert_X[y==0, 0], revert_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(revert_X[y==1, 0], revert_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *revertedV[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *revertedV[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "inverse  = np.linalg.inv(vec)\n",
    "inverse  = inverse - inverse*inverse.std(0,keepdims=True)\n",
    "revert_X   = project_X @ inverse ;revertedV  = project_V @ inverse.T\n",
    "\n",
    "plt.subplot(144); \n",
    "plt.scatter(revert_X[y==0, 0], revert_X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(revert_X[y==1, 0], revert_X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *revertedV[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *revertedV[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# kernel PCA\n",
    "gamma = 100\n",
    "sq_x  = -2 * X @ X.T + np.sum(X**2,1) + np.sum(X**2,1,keepdims=True)\n",
    "K     = np.exp(-gamma * sq_x)\n",
    "N     = K.shape[0]\n",
    "one_n = np.ones((N,N)) / N\n",
    "K2    = K - one_n.dot(K) - K.dot(one_n) + one_n.dot(K).dot(one_n)\n",
    "eigvals, eigvecs = np.linalg.eigh(K2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_pc = eigvecs[:,-2:]\n",
    "\n",
    "plt.figure(figsize=(15,5)); \n",
    "\n",
    "plt.subplot(121); \n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "new_pc_cen = new_pc - new_pc.mean(0,keepdims=True)\n",
    "cov        = new_pc_cen.T @ new_pc_cen /(new_pc_cen.shape[0] - 1)\n",
    "val,vec    = np.linalg.eigh(cov)\n",
    "\n",
    "plt.subplot(122); \n",
    "plt.scatter(new_pc[y==0, 0], new_pc[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(new_pc[y==1, 0], new_pc[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.005, head_length=0.005,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.005, head_length=0.005,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     3,
     5
    ]
   },
   "outputs": [],
   "source": [
    "# Spase PCA\n",
    "import pywt\n",
    "\n",
    "class SPC(object):\n",
    "    \n",
    "    def __init__(self, number_of_components,max_iter=10, threshold_val=1.5 ):\n",
    "        \n",
    "        \n",
    "        \"\"\"Initialize the SPC object\n",
    "        \n",
    "        Positional arguments:\n",
    "        number_of_components -- the number of sparse principal components \n",
    "        to compute, must be between 1 and p (total number of features)\n",
    "        \n",
    "        Keyword argument:\n",
    "        max_iter -- the number of iterations to perform (default=10)\n",
    "        threshold_val -- value of the lambda regularisation \n",
    "        parameter (default=10)\n",
    "        \"\"\"\n",
    "        self.number_of_components=number_of_components\n",
    "        self.max_iter=max_iter\n",
    "        self.threshold_val=threshold_val\n",
    "    \n",
    "    def fit(self, X_):\n",
    "        \"\"\"learn the sparse pc of a data matrix, return sparse estimates\n",
    "        of the left and right singular vectors (U and V respectively) \n",
    "        as well as the standard principal components loading matrix W\n",
    "    \n",
    "        Positional arguments:\n",
    "        X_ -- training data matrix, as numpy ndarray\n",
    "        \n",
    "        \"\"\" \n",
    "        print(\"computing sparse principal components...\")\n",
    "        print(\"computing SVD of data matrix...\")\n",
    "        U, s, V = np.linalg.svd(X_, full_matrices=True)  \n",
    "        cnt = 0\n",
    "        self.U = U\n",
    "        self.W = V.T\n",
    "        def normalize(vector):\n",
    "            norm=np.linalg.norm(vector)\n",
    "            if norm>0:\n",
    "                return vector/norm\n",
    "            else:\n",
    "                return vector\n",
    "        print(\"starting iterations...\")\n",
    "        while True:\n",
    "           \n",
    "            self.V = pywt.threshold(np.dot(U[:self.number_of_components],X_), self.threshold_val)\n",
    "            self.U = np.dot(self.V,X_.T)\n",
    "            self.U = np.array([normalize(u_i) for u_i in self.U])\n",
    "            if cnt%2==0:\n",
    "                print(\"{} out of {} iterations\".format(cnt,self.max_iter))\n",
    "            cnt += 1\n",
    "            if cnt == self.max_iter:\n",
    "                self.V = np.array([normalize(v_i) for v_i in self.V])\n",
    "                break\n",
    "        print(\"...finish\")\n",
    "        return self.U, self.V, self.W\n",
    "    \n",
    "    def transform(self, X_, k=2):\n",
    "        X_reduced_spca     = np.dot(X_, np.dot(self.V[:k].T, self.V[:k]))\n",
    "        return X_reduced_spca\n",
    "my_spca  = SPC(2,3000,0.1)\n",
    "my_spca.fit(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparse_data = my_spca.transform(X)\n",
    "\n",
    "plt.figure(figsize=(15,5)); \n",
    "\n",
    "plt.subplot(121); \n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "new_pc_cen = sparse_data - sparse_data.mean(0,keepdims=True)\n",
    "cov        = new_pc_cen.T @ new_pc_cen /(new_pc_cen.shape[0] - 1)\n",
    "val,vec    = np.linalg.eigh(cov)\n",
    "\n",
    "plt.subplot(122); \n",
    "plt.scatter(new_pc[y==0, 0], new_pc[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(new_pc[y==1, 0], new_pc[y==1, 1], color='blue', alpha=0.5)\n",
    "pca21=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.005, head_length=0.005,color='Green',  label='First PC')\n",
    "pca22=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.005, head_length=0.005,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     76
    ]
   },
   "outputs": [],
   "source": [
    "# robust pca\n",
    "from __future__ import division, print_function\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "try:\n",
    "    from pylab import plt\n",
    "except ImportError:\n",
    "    print('Unable to import pylab. R_pca.plot_fit() will not work.')\n",
    "\n",
    "try:\n",
    "    # Python 2: 'xrange' is the iterative version\n",
    "    range = xrange\n",
    "except NameError:\n",
    "    # Python 3: 'range' is iterative - no need for 'xrange'\n",
    "    pass\n",
    "\n",
    "\n",
    "class R_pca:\n",
    "\n",
    "    def __init__(self, D, mu=None, lmbda=None):\n",
    "        self.D = D\n",
    "        self.S = np.zeros(self.D.shape)\n",
    "        self.Y = np.zeros(self.D.shape)\n",
    "\n",
    "        if mu:\n",
    "            self.mu = mu\n",
    "        else:\n",
    "            self.mu = np.prod(self.D.shape) / (4 * self.norm_p(self.D, 2))\n",
    "\n",
    "        self.mu_inv = 1 / self.mu\n",
    "\n",
    "        if lmbda:\n",
    "            self.lmbda = lmbda\n",
    "        else:\n",
    "            self.lmbda = 1 / np.sqrt(np.max(self.D.shape))\n",
    "\n",
    "    @staticmethod\n",
    "    def norm_p(M, p):\n",
    "        return np.sum(np.power(M, p))\n",
    "\n",
    "    @staticmethod\n",
    "    def shrink(M, tau):\n",
    "        return np.sign(M) * np.maximum((np.abs(M) - tau), np.zeros(M.shape))\n",
    "\n",
    "    def svd_threshold(self, M, tau):\n",
    "        U, S, V = np.linalg.svd(M, full_matrices=False)\n",
    "        return np.dot(U, np.dot(np.diag(self.shrink(S, tau)), V))\n",
    "\n",
    "    def fit(self, tol=None, max_iter=1000, iter_print=100):\n",
    "        iter = 0\n",
    "        err = np.Inf\n",
    "        Sk = self.S\n",
    "        Yk = self.Y\n",
    "        Lk = np.zeros(self.D.shape)\n",
    "\n",
    "        if tol:\n",
    "            _tol = tol\n",
    "        else:\n",
    "            _tol = 1E-7 * self.norm_p(np.abs(self.D), 2)\n",
    "\n",
    "        while (err > _tol) and iter < max_iter:\n",
    "            Lk = self.svd_threshold(\n",
    "                self.D - Sk + self.mu_inv * Yk, self.mu_inv)\n",
    "            Sk = self.shrink(\n",
    "                self.D - Lk + (self.mu_inv * Yk), self.mu_inv * self.lmbda)\n",
    "            Yk = Yk + self.mu * (self.D - Lk - Sk)\n",
    "            err = self.norm_p(np.abs(self.D - Lk - Sk), 2)\n",
    "            iter += 1\n",
    "            if (iter % iter_print) == 0 or iter == 1 or iter > max_iter or err <= _tol:\n",
    "                print('iteration: {0}, error: {1}'.format(iter, err))\n",
    "\n",
    "        self.L = Lk\n",
    "        self.S = Sk\n",
    "        return Lk, Sk\n",
    "\n",
    "    def plot_fit(self, size=None, tol=0.1, axis_on=True):\n",
    "\n",
    "        n, d = self.D.shape\n",
    "\n",
    "        if size:\n",
    "            nrows, ncols = size\n",
    "        else:\n",
    "            sq = np.ceil(np.sqrt(n))\n",
    "            nrows = int(sq)\n",
    "            ncols = int(sq)\n",
    "\n",
    "        ymin = np.nanmin(self.D)\n",
    "        ymax = np.nanmax(self.D)\n",
    "        print('ymin: {0}, ymax: {1}'.format(ymin, ymax))\n",
    "\n",
    "        numplots = np.min([n, nrows * ncols])\n",
    "        plt.figure()\n",
    "\n",
    "        for n in range(numplots):\n",
    "            plt.subplot(nrows, ncols, n + 1)\n",
    "            plt.ylim((ymin - tol, ymax + tol))\n",
    "            plt.plot(self.L[n, :] + self.S[n, :], 'r')\n",
    "            plt.plot(self.L[n, :], 'b')\n",
    "            if not axis_on:\n",
    "                plt.axis('off')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rpca = R_pca(X)\n",
    "L, S = rpca.fit(max_iter=10000, iter_print=1)\n",
    "print(L.shape)\n",
    "print(S.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(15,5)); \n",
    "\n",
    "plt.subplot(131); \n",
    "plt.scatter(X[y==0, 0], X[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(X[y==1, 0], X[y==1, 1], color='blue', alpha=0.5)\n",
    "pca11=plt.arrow   (0, 0, *vec[:,0] * val[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "pca12=plt.arrow   (0, 0, *vec[:,1] * val[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(132); \n",
    "plt.scatter(L[y==0, 0], L[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(L[y==1, 0], L[y==1, 1], color='blue', alpha=0.5)\n",
    "# pca21=plt.arrow   (0, 0, *project_V[:,0] * val_reduced[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "# pca22=plt.arrow   (0, 0, *project_V[:,1] * val_reduced[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.subplot(133); \n",
    "plt.scatter(S[y==0, 0], S[y==0, 1], color='red',  alpha=0.5)\n",
    "plt.scatter(S[y==1, 0], S[y==1, 1], color='blue', alpha=0.5)\n",
    "# pca21=plt.arrow   (0, 0, *revertedV[:,0] * val_reduced[0], head_width=0.05, head_length=0.05,color='Green',  label='First PC')\n",
    "# pca22=plt.arrow   (0, 0, *revertedV[:,1] * val_reduced[1], head_width=0.05, head_length=0.05,color='magenta',label='Second PC')\n",
    "plt.grid(True); \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference\n",
    "1. tf.set_random_seed | TensorFlow. (2018). TensorFlow. Retrieved 28 October 2018, from https://www.tensorflow.org/api_docs/python/tf/set_random_seed\n",
    "2. Comparing different hierarchical linkage methods on toy datasets — scikit-learn 0.20.0 documentation. (2018). Scikit-learn.org. Retrieved 28 October 2018, from http://scikit-learn.org/stable/auto_examples/cluster/plot_linkage_comparison.html#sphx-glr-auto-examples-cluster-plot-linkage-comparison-py\n",
    "3. VanderPlas, J. (2018). Multiple Subplots | Python Data Science Handbook. Jakevdp.github.io. Retrieved 28 October 2018, from https://jakevdp.github.io/PythonDataScienceHandbook/04.08-multiple-subplots.html\n",
    "4. object, M. (2018). Matplotlib returning a plot object. Stack Overflow. Retrieved 28 October 2018, from https://stackoverflow.com/questions/43925337/matplotlib-returning-a-plot-object\n",
    "5. Plot randomly generated classification dataset — scikit-learn 0.20.0 documentation. (2018). Scikit-learn.org. Retrieved 28 October 2018, from http://scikit-learn.org/stable/auto_examples/datasets/plot_random_dataset.html#sphx-glr-auto-examples-datasets-plot-random-dataset-py\n",
    "6. 5. Dataset loading utilities — scikit-learn 0.20.0 documentation. (2018). Scikit-learn.org. Retrieved 28 October 2018, from http://scikit-learn.org/stable/datasets/index.html\n",
    "7. mplot3d tutorial — Matplotlib 2.0.2 documentation. (2018). Matplotlib.org. Retrieved 28 October 2018, from https://matplotlib.org/mpl_toolkits/mplot3d/tutorial.html\n",
    "8. matplotlib, H. (2018). How to plot vectors in python using matplotlib. Stack Overflow. Retrieved 28 October 2018, from https://stackoverflow.com/questions/42281966/how-to-plot-vectors-in-python-using-matplotlib\n",
    "9. arrow, M. (2018). Matplotlib legend for an arrow. Stack Overflow. Retrieved 28 October 2018, from https://stackoverflow.com/questions/22348229/matplotlib-legend-for-an-arrow\n",
    "10. plt colors - Google Search. (2018). Google.ca. Retrieved 28 October 2018, from https://www.google.ca/search?q=plt+colors&rlz=1C1CHBF_enCA771CA771&oq=plt+colors&aqs=chrome..69i57j0l5.1038j0j7&sourceid=chrome&ie=UTF-8\n",
    "11. Brownlee, J. (2018). A Gentle Introduction to Singular-Value Decomposition for Machine Learning. Machine Learning Mastery. Retrieved 28 October 2018, from https://machinelearningmastery.com/singular-value-decomposition-for-machine-learning/\n",
    "12. https://www.quora.com/Why-do-we-need-to-center-the-data-for-Principle-Components-Analysis\n",
    "13. Distance Matrix Vectorization Trick – Manifold Blog – Medium. (2016). Medium. Retrieved 29 October 2018, from https://medium.com/dataholiks-distillery/l2-distance-matrix-vectorization-trick-26aa3247ac6c\n",
    "14. (pdist), H. (2018). How does condensed distance matrix work? (pdist). Stack Overflow. Retrieved 29 October 2018, from https://stackoverflow.com/questions/13079563/how-does-condensed-distance-matrix-work-pdist\n",
    "15. aboyker/easyspc. (2018). GitHub. Retrieved 29 October 2018, from https://github.com/aboyker/easyspc/blob/master/build/lib/easyspc/__init__.py\n",
    "16. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
