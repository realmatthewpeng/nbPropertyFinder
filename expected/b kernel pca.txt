X    = X - X.mean(0,keepdims=True)